nutils/function.py

```diff
@@ -124,15 +124,18 @@ def __init__(self, shape: Shape, dtype: DType) -> None:
         # Try to convert `length` to an `int` by lowering to an `Evaluable` and
         # evaluating.
         try:
-          length = int(length.lower().eval())
+          length = int(length.as_evaluable_array().eval())
         except:
           pass
       shape_.append(length)
     self.shape = tuple(shape_)
     self.dtype = dtype
 
   def as_evaluable_array(self) -> evaluable.Array:
-    return self.lower()
+    result = getattr(self, '_as_evaluable_array', None)
+    if result is None:
+      self._as_evaluable_array = result = self.lower()
+    return result
```

Comment:
If we make this a property we could use `cached_property`.

Reply:
done